packages/backend/src/shared/core/Result.ts

```diff
@@ -0,0 +1,91 @@
+export class Result<T> {
+  public isSuccess: boolean;
+  public isFailure: boolean;
+  public error?: T | string;
+  private _value?: T;
+
+  public constructor(isSuccess: boolean, error?: T | string, value?: T) {
+    if (isSuccess && error) {
+      throw new Error('InvalidOperation: A result cannot be successful and contain an error');
+    }
+    if (!isSuccess && !error) {
+      throw new Error('InvalidOperation: A failing result needs to contain an error message');
+    }
+
+    this.isSuccess = isSuccess;
+    this.isFailure = !isSuccess;
+    this.error = error;
+    this._value = value;
+
+    Object.freeze(this);
+  }
+
+  public getValue(): T {
+    if (!this.isSuccess || !this._value) {
+      throw new Error("Can't get the value of an error result. Use 'errorValue' instead.");
+    }
+
+    return this._value;
+  }
+
+  public errorValue(): T {
+    return this.error as T;
+  }
+
+  public static ok<U>(value?: U): Result<U> {
+    return new Result<U>(true, undefined, value);
+  }
+
+  public static fail<U>(error: string | U): Result<U> {
+    return new Result<U>(false, error);
+  }
+
+  public static combine(results: Result<any>[]): Result<any> {
+    for (const result of results) {
+      if (result.isFailure) return result;
+    }
+    return Result.ok();
+  }
+}
+
+export type Either<L, A> = Left<L, A> | Right<L, A>;
+
+export class Left<L, A> {
+  readonly value: L;
+
+  constructor(value: L) {
+    this.value = value;
+  }
+
+  isLeft(): this is Left<L, A> {
+    return true;
+  }
+
+  isRight(): this is Right<L, A> {
+    return false;
+  }
+}
+
+export class Right<L, A> {
+  readonly value: A;
+
+  constructor(value: A) {
+    this.value = value;
+  }
+
+  isLeft(): this is Left<L, A> {
+    return false;
+  }
+
+  isRight(): this is Right<L, A> {
+    return true;
+  }
+}
+
+export const left = <L, A>(l: L): Either<L, A> => {
+  return new Left(l);
```

Comment:
don't you need to pass types here as well as you did below?
```suggestion
  return new Left<L, A>(l);
```